src/lib.rs

```diff
@@ -1 +1,192 @@
 #![allow(non_snake_case)]
+
+use std::ffi::{c_void, OsStr, OsString};
+use std::mem::size_of;
+use std::path::{Path, PathBuf};
+use thiserror::Error;
+use windows::Win32::Foundation::{CloseHandle, GetLastError, PWSTR};
+use windows::Win32::Security::SECURITY_ATTRIBUTES;
+use windows::Win32::System::Threading::{
+    GetExitCodeProcess, TerminateProcess, WaitForSingleObject, PROCESS_CREATION_FLAGS,
+    PROCESS_INFORMATION, STARTUPINFOW,
+};
+use windows::Win32::System::WindowsProgramming::INFINITE;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("cannot create process: {0}")]
+    CreationFailed(u32),
+    #[error("cannot wait process: {0}")]
+    WaitFailed(u32),
+    #[error("cannot kill process: {0}")]
+    KillFailed(u32),
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug)]
+pub struct Command {
+    command: OsString,
+    inherit_handles: Option<bool>,
+    current_directory: Option<PathBuf>,
+}
+
+impl Command {
+    pub fn new(command: impl AsRef<OsStr>) -> Command {
```

Comment:
Oh this is nice! You took it from std::command I assume?

I'm not sure why they used `AsRef` here as they need to own the type in the end... :thinking: :interrobang: In your case you're storing OsString so you should use `impl Into<OsString>` instead to show your intention to clone it.

In their case it looks like they do something a bite more optimized:

1. [That's their constructor.](https://github.com/rust-lang/rust/blob/848a38ac9d9bc956cded893bdd55e1171a2706dd/library/std/src/sys/unix/process/process_common.rs#L133) It's calling os2c on the reference.
2. [That's os2c.](https://github.com/rust-lang/rust/blob/848a38ac9d9bc956cded893bdd55e1171a2706dd/library/std/src/sys/unix/process/process_common.rs#L327) It's instantiating a CString instead of an OsString. CString are [null-terminated](https://en.wikipedia.org/wiki/Null-terminated_string) which apparently is required to make the actual call in the kernel.

So I guess this is a way to avoid an extra allocation:
- they do `AsRef<OsStr>` -> CString
- instead of doing `Into<OsString>` -> OsString -> CString

Pretty sure the Windows' strings are also terminated though. I'll will check if we can make this better but in the meantime you can replace `AsRef<OsStr>` to `Into<OsString>` it will make your code simpler and more correct.

Reply:
I found what the automatic conversion `IntoParam<PWSTR>` does [here](https://github.com/microsoft/windows-rs/blob/1801b69846d5d1faaa33a0d289f5d5c2e0012825/src/runtime/bindings.rs#L1598). To do that, I download the sources and I grepped like this: `grep -r 'IntoParam.\+PWSTR' src/`

The Interesting part is this one: `self.encode_wide().chain(::core::iter::once(0)).collect::<std::vec::Vec<u16>>()`. Lets deconstruct this:
1. `self.encode_wide()`: [doc](https://doc.rust-lang.org/std/os/windows/ffi/trait.OsStrExt.html#tymethod.encode_wide). re-encode the input
2. `.chain(::core::iter::once(0))`: add a 0 (nul) at the end of the iterator
3. `.collect::<std::vec::Vec<u16>>()`: allocates memory and transform the iterator to a vec.

This means using OsString is not optimal because it still requires an extra step of conversion: OsString -> PWSTR. (Same story for String)

If we want to optimize the allocations here, we should use the same type than them in argument: [`IntoParam<'a, PWSTR>`](https://github.com/microsoft/windows-rs/blob/6e1a214540654da5a9775df11458979e9ee9d263/src/runtime/into_param.rs#L4).

But as you can see in the code, they added `#[doc(hidden)]` on top of it. This means it's public-ish but they don't want us to use this and it is not part of their standard API. So there is nothing we can do.

Only 2 choices left: either use String either OsString. OsString is the better choice here because [String can always be converted to OsString](https://doc.rust-lang.org/std/string/struct.String.html#impl-From%3CString%3E-7) while the other way around is not true: [OsString cannot always be converted to String](https://doc.rust-lang.org/std/ffi/struct.OsString.html#method.into_string).

Reply:
Yeah because OsString are not UTF-8 compliant.
TIL about `#[doc(hidden)]`